news/2020-09-15-CDI_Lite_extension.adoc

```diff
@@ -0,0 +1,284 @@
+
+When we started thinking about a lite API design space, we mostly constrained ourselves to one particularly common use case: annotation transformations. Many things that Portable Extensions allow can be achieved simply by adding an annotation here or removing an annotation there. We know that other transformations are necessary, but the API for annotation transformation is the most developed proposal we have. The other ones are significantly more rough, so bear with us please and submit feedback! (*TODO* where?)
+
+Before we start, we’d also like to note that in this text, we assume the reader (that is, you!) is familiar with CDI and preferably also Portable Extensions, as we expect some knowledge and don’t explain everything. We also make references to Portable Extensions on several occasions.
+
+== Extensions
+
+In our proposal, Build Compatible Extensions are simply methods annotated with an _extension annotation_. Extension annotations correspond to phases in which extensions are processed. There are 4 phases:
+
+. *@Discovery*: adding classes to application, registering custom contexts
+. *@Enhancement*: transforming annotations
+. *@Synthesis*: registering synthetic beans and observers
+. *@Validation*: performing custom validation
+
+There are some constraints we put on these methods (such as: they must be `public`), but they should be pretty obvious and shouldn’t be limiting anyone. The container automatically finds extensions and invokes them when time is right.
+When exactly are the extensions invoked can’t be defined in too much detail, because we want implementations to be able to invoke them at build time (e.g. during application compilation) or runtime (e.g. during application deployment). It certainly happens before the application starts. Extensions in earlier phases are guaranteed to run before extensions in later phases.
+These methods can declare an arbitrary number of parameters, all of which are supplied by the container. There’s a set of predefined types of parameters for each phase, and all implementations would have to support that. We’re also thinking of designing an SPI that would allow anyone to contribute support for other types of parameters.
+A class can declare multiple extensions, in which case they are all invoked on a single instance of the class. If you need to control order of extension invocations, there’s an annotation *`@ExtensionPriority`* just for that.
+This is a lot of text already, so let’s take a look at an example:
+
+[source,java]
+----
+public class MyExtension {
+    @Enhancement
+    public void doSomething() {
+        System.out.println("This is an extension, yay!");
+    }
+}
+----
+
+This doesn’t really do anything, just prints a message whenever the extension is executed. Let’s create something more interesting. Say, moving a qualifier annotation from one class to another. Let’s assume that we have these classes in our application:
+
+* *A qualifier annotation*
```

Comment:
Maybe we should rewrite the list to a section with explicit number in the title, i.e. `===1. A qualifier annotation`. Alternatively, we could use a numbered list [with continuation](https://asciidoctor.org/docs/user-manual/#list-continuation), i.e.
```
1. A qualifier annotation
+
[source,java]
----
...
----
```

Reply:
I agree the current state is hard to read. I'd suggest using `====` (subtitle format, basically just a little bigger text)

Reply:
Agree that numbered list wasn't the best choice, feel free to reformat as you wish :-)